### PR TITLE
Validator should only check for units by name

### LINKED ISSUE #715: Should imported units be "found" by the validator?
```
TEST(ImportSource, KRM)
{
    auto model = libcellml::Model::create("model");
    auto component = libcellml::Component::create("component");
    auto importSource = libcellml::ImportSource::create();
    importSource->setUrl("somewhere.cellml");
    model->addImportSource(importSource);
    model->addComponent(component);

    auto mV = libcellml::Units::create("mV");
    mV->setImportReference("mV");
    mV->setImportSource(importSource); // << removing this means that the model is valid ... 
    model->addUnits(mV);

    auto v = libcellml::Variable::create("variable");
    v->setUnits(mV);
    component->addVariable(v);

    EXPECT_EQ(model, mV->parent());
    EXPECT_TRUE(model->hasUnits("mV"));

    auto validator = libcellml::Validator::create();
    validator->validateModel(model);
    printIssues(validator);  // << this says it can't find the units ... :(
}
```

### PR DESCRIPTION
**Pull request by kerimoyle**
Addresses #715

### CODE REVIEW

File: src/validator.cpp [hunk 1]
```diff
@@ -635,7 +635,7 @@ void Validator::ValidatorImpl::validateVariable(const VariablePtr &variable, con
     } else if (!isStandardUnitName(unitsName)) {
         ComponentPtr component = owningComponent(variable);
         ModelPtr model = owningModel(component);
-        if ((model != nullptr) && !model->hasUnits(variable->units())) {
+        if ((model != nullptr) && !model->hasUnits(unitsName)) {
```
Comment (agarny) [inline src/validator.cpp:688]: This clearly changes everything and, looking at `Model::ModelImpl::findUnits()` (which is the method effectively used here), we can indeed see that the implementation is completely different depending on whether we pass it a `Units` or a string. I find this very weird and I am not sure I like it.

File: tests/validator/validator.cpp [hunk 2]
```diff
@@ -2625,10 +2625,36 @@ TEST(Validator, refToUnitsByNameNeedsLinkUnitsToValidate)
     t2->setUnits("millisecond");
 
     validator->validateModel(model);
-    EXPECT_EQ(size_t(1), validator->errorCount());
+    EXPECT_EQ(size_t(1), validator->errorCount()); // KRM this will fail now since we *don't* need to link units anymore
```
Comment (agarny) [inline tests/validator/validator.cpp]: I guess this test will become moot if we were to go with your proposed change?

Comment (kerimoyle) [reply tests/validator/validator.cpp]: Exactly - that's why I'm asking.

File: src/validator.cpp (hunk 1 above)
Comment (kerimoyle) [reply src/validator.cpp:688]: I think it's ok to have a `hasUnits` function that also tests content fo the units; just not that it's used here as it doesn't come into validation.

**Review by hsorby — DISMISSED** [review]

**Comment by hsorby** [comment]
I think the `Validator` should only be checking that the reference is valid.

**Review by hsorby — CHANGES_REQUESTED** [review]
I think the implementation is correct but the tests need to change to show the new understanding.

**Comment by kerimoyle** [comment]
> I think the implementation is correct but the tests need to change to show the new understanding.

Cool, ta.  I'll remove that test and resubmit - it's not needed any more.

**Comment by kerimoyle** [comment]
I've had to make a couple of other tweaks too and added a couple of small tests.  Ready for review please @agarny @hsorby @nickerso

File: src/validator.cpp (hunk 1 above)
Comment (agarny) [reply src/validator.cpp:688]: Maybe, but when you have a method with two different signatures, I would still expect that method to work the same while here `hasUnits()` does different things (through `findUnits()`) and that's what I find dodgy.

**Review by agarny — DISMISSED** [review]
Apart from my general comment about `hasUnits()`, it looks good to me.

**Comment by kerimoyle** [comment]
> Apart from my general comment about hasUnits(), it looks good to me.
@agarny I do take your point ... but even though that's been highlighted by this PR, it's not something that this PR is meant to address.  I'm just calling a different function in the validator, not changing how we search for anything.  Can we make that discussion a separate issue?

**Comment by agarny** [comment]
> > Apart from my general comment about hasUnits(), it looks good to me.
> > @agarny I do take your point ... but even though that's been highlighted by this PR, it's not something that this PR is meant to address.  I'm just calling a different function in the validator, not changing how we search for anything.  Can we make that discussion a separate issue?

Agreed and that's exactly why I still approved your PR despite my concern about the different implementations of `hasUnits()`, for which I have just created an issue (see issue #719).

**Comment by kerimoyle** [comment]
Great, thanks @agarny :)

File: src/units.cpp [hunk 3]
```diff
@@ -453,18 +456,17 @@ double Units::scalingFactor(const UnitsPtr &units1, const UnitsPtr &units2, bool
     bool updateUnits2 = false;
 
     if ((units1 != nullptr) && (units2 != nullptr)) {
-        double multiplier = 0.0;
+        if ((owningModel(units1) == owningModel(units2)) && (units1->name() == units2->name())) {
+            return 1.0;
+        }
```
Comment (hsorby) [inline src/units.cpp]: I don't understand this addition, why should I care about who owns the units.  What does it have to do with the validator only checking units by name.

File: tests/units/units.cpp [hunk 4]
```diff
@@ -2469,3 +2469,52 @@ TEST(Units, addUnitsMultipleTimes)
...
+    u2->addUnit("second");
+    u1->addUnit("second", "milli");
+
+    auto scaling = libcellml::Units::scalingFactor(u1, u5);
+    EXPECT_EQ(0.0, scaling);
+}
+
+TEST(Units, scalingFactorBetweenSameUnits)
+{
+    auto model = libcellml::Model::create("model");
```
Comment (hsorby) [inline tests/units/units.cpp:2522]: What is model doing here?

File: tests/units/units.cpp [hunk 5]
```diff
@@ -2469,3 +2469,52 @@ TEST(Units, addUnitsMultipleTimes)
...
+    auto u1 = libcellml::Units::create("units");
+    u1->addUnit("second", "milli");
+    model1->addUnits(u1);
+
+    auto model2 = model1->clone();
+    auto u2 = model2->units(0);
+
+    auto scaling = libcellml::Units::scalingFactor(u1, u2);
+    EXPECT_EQ(1.0, scaling);
+}
```
Comment (hsorby) [inline tests/units/units.cpp:2544]: Where is the co-test for this a non-unit scaling factor for units of different scale?

**Review by hsorby — DISMISSED** [review]

File: src/units.cpp (hunk 3 above)
Comment (kerimoyle) [reply src/units.cpp]: It's perfectly valid to have two units in separate models with the same name.  There was a check lower down the function that returned a scaling factor of 1 if the two units had the same name.  That's only "safe" (ish) when they're from the same model.  It's not directly related to the validation stuff, but when I changed that another test broke, so this is making sure that all our current tests continue to pass even with the change.

File: tests/units/units.cpp (hunk 4 above)
Comment (kerimoyle) [reply tests/units/units.cpp:2522]: See line 196 in src/units.cpp.  In order to check for child unit references, we need to have access to the other units in the model... so it needs a model.

File: tests/units/units.cpp (hunk 5 above)
Comment (kerimoyle) [reply tests/units/units.cpp:2544]: There are loads of tests for scaling factors earlier in this file - I don't quite understand which aspect is missing?  This test was intended to show that even though these two have the same name, they're still tested properly.  I guess changing the  scale away from 1 would show that more clearly - is that what you meant?

File: src/units.cpp (hunk 3 above)
Comment (hsorby) [reply src/units.cpp]: It was further down for a reason right.  After all else is said and done if the units have the same name then they are considered to have a scaling factor of `1.0` given this is the case by definition.

File: tests/units/units.cpp (hunk 5 above)
Comment (kerimoyle) [reply tests/units/units.cpp:2544]: 6008b53

File: tests/units/units.cpp (hunk 4 above)
Comment (hsorby) [reply tests/units/units.cpp:2522]: But `model` is not used in this test?

File: src/units.cpp (hunk 3 above)
Comment (kerimoyle) [reply src/units.cpp]: I'm not understanding two parts of that:
> It was further down for a reason right.

There is no change to the functionality at all - the only thing that's different is that we can skip a whole lot of recursive calculation if we are going to return a value based on the name in any case.  I don't understand how doing those extra calculations is a good reason to have that test further down the function?

> After all else is said and done if the units have the same name then they are considered to have a scaling factor of 1.0 given this is the case by definition.

I don't agree.  Why would they be considered the same? We're checking units definitions here?

File: tests/units/units.cpp (hunk 4 above)
Comment (kerimoyle) [reply tests/units/units.cpp:2522]: Duh! I get it now.  Changed in a8c89fdb

**Comment by kerimoyle** [comment]
Ready for re-review please @agarny @hsorby @nickerso

**Review by agarny — DISMISSED** [review]

**Comment by agarny** [comment]
Ping @hsorby and @nickerso for re-reviewing this PR.

File: src/units.cpp (hunk 3 above)
Comment (nickerso) [reply src/units.cpp]: I think this is still an unresolved issue and this change breaks the way this method should work

**Review by agarny — DISMISSED** [review]

**Review by nickerso — DISMISSED** [review]

**Review by agarny — APPROVED** [review]

**Review by nickerso — APPROVED** [review]

**Pull request merged**